src/Rings/msolve/interface.jl

```diff
@@ -0,0 +1,160 @@
+            cc += 1
+        end
+        for e in Singular.exponent_vectors(id[i])
+            for j = 1:nvars
+                exps[nvars*ec+j]  =  Base.Int(e[j])
+            end
+            ec +=  1
+        end
+    end
+    lens, cfs, exps
+end
+
+function convert_qq_singular_ideal_to_array(
+        id::Singular.sideal
+        )
+    ngens   = Singular.ngens(id)
+    nvars   = Singular.nvars(base_ring(id))
+    nterms  = 0
+    lens = Array{Int32,1}(undef, ngens)
+    for i = 1:ngens
+        lens[i] =   Singular.length(id[i])
+        nterms  +=  lens[i]
+    end
+    cfs   = Array{BigInt,1}(undef, 2*nterms)
+    exps  = Array{Int32,1}(undef, nvars*nterms)
+    cc = 1 # coefficient counter
+    ec = 0 # exponent vector counter
+    for i = 1:Singular.ngens(id)
+        for c in Singular.coeffs(id[i])
+            cfs[cc] = Singular.libSingular.n_GetMPZ(numerator(c).ptr, Singular.ZZ.ptr)
+            cc += 1
+            cfs[cc] = Singular.libSingular.n_GetMPZ(denominator(c).ptr, Singular.ZZ.ptr)
+            cc += 1
+        end
+        for e in Singular.exponent_vectors(id[i])
+            for j = 1:nvars
+                exps[nvars*ec+j]  =  Base.Int(e[j])
+            end
+            ec +=  1
+        end
+    end
+    lens, cfs, exps
+end
+
+# we know that the terms are already sorted and they are all different
+# coming from GroebnerBasis' F4 computation, so we do not need p_Add_q for
+# the terms, but we can directly set the next pointers of the polynomials
+function convert_ff_gb_array_to_singular_ideal(
+        bld::Int32,
+        blen::Array{Int32,1},
+        bexp::Array{Int32,1},
+        bcf::Array{Int32,1},
+        R::Singular.PolyRing
+        )
+    ngens = bld
+
+    nvars = Singular.nvars(R)
+    basis = Singular.Ideal(R, ) # empty ideal
+    # first entry in exponent vector is module component => nvars+1
+    exp   = zeros(Cint, nvars+1)
+
+    list  = Singular.elem_type(R)[]
+    # we generate the singular polynomials low level in order
+    # to avoid overhead due to many exponent operations etc.
+    j   = ngens + 1 + 1
+    len = 1
+    for i = 1:ngens
+        # do the first term
+        p = Singular.libSingular.p_Init(R.ptr)
+        Singular.libSingular.pSetCoeff0(p, Clong(bcf[len]), R.ptr)
+        for k = 1:nvars
+            exp[k+1]  = bexp[(len-1) * nvars + k]
+        end
+        Singular.libSingular.p_SetExpV(p, exp, R.ptr)
+        lp  = p
+        for j = 2:blen[i]
+          pterm = Singular.libSingular.p_Init(R.ptr)
+          Singular.libSingular.pSetCoeff0(pterm, Clong(bcf[len+j-1]), R.ptr)
+          for k = 1:nvars
+              exp[k+1]  = bexp[(len+j-1-1) * nvars + k]
+          end
+          Singular.libSingular.p_SetExpV(pterm, exp, R.ptr)
+          Singular.libSingular.SetpNext(lp, pterm)
+          lp  = pterm
+        end
+        push!(list, R(p))
+        len += blen[i]
+    end
+    return Singular.Ideal(R, list)
+end
+
+function convert_qq_gb_array_to_singular_ideal(
```

Comment:
This function is apparently not covered by the test right now, is that intentional?

Reply:
I unified the conversion function from singular ideal to msolve input as suggested. Moreover, I removed, for the time being, `convert_qq_gb_array_to_singular_ideal`: It is not used right now, but it will be necessary once the next release of `msolve` is available. Maybe I can then also unify the `_ff_` and the `_qq_` variants.

Reply:
Thank you!